app/pages/learn/01_tutorial/04_mastering-the-api/02_invoke/00_methodhandle.md

```diff
@@ -0,0 +1,531 @@
+```
+
+It works similarly for the `Field` object:
+
+```java
+Field field = MethodHandles.reflectAs(Field.class, getterMethodHandle); // same result is achieved by reflecting `setterMethodHandle`
+```
+
+<a id="exceptions">&nbsp;</a>
+## Exception handling
+
+Both `invokeExact` and `invoke` throw `Throwable`, so there is no limitation to what an underlying method can throw. 
+The method that invokes a method handle must either explicitly throw a `Throwable` or catch it.
+
+There are certain methods in the `MethodHandles` API that can make exception handling easier. 
+Let's take a look at several examples.
+
+### `catch` wrapper
+`MethodHandles.catchException` method can wrap a given method handle inside a provided exception handler method handle.
+
+Say, we have a method `problematicMethod` that does some job, and a method `exceptionHandler` that handles a particular exception `IllegalArgumentException`. 
+The exception handler method must return the same type as the original method. 
+The first argument it accepts is a `Throwable` that we're interested in, after which follow the rest of the arguments that we've originally accepted:
+
+```java
+public static int problematicMethod(String argument) throws IllegalArgumentException {
+    if ("invalid".equals(argument)) {
+        throw new IllegalArgumentException();
+    }
+    return 1;
+}
+
+public static int exceptionHandler(IllegalArgumentException e, String argument) {
+    // log exception
+    return 0;
+}
+```
+
+We can look up the method handles for both these methods and wrap `problematicMethod` inside an `exceptionHandler`. 
+The resulting `MethodHandle` on invocation will handle the `IllegalArgumentException` properly, continuing to throw any other exceptions if they arise.
+
+```java
+MethodHandles.Lookup lookup = MethodHandles.lookup();
+MethodHandle methodHandle = lookup.findStatic(Example.class, "problematicMethod", MethodType.methodType(int.class, String.class));
+MethodHandle handler = lookup.findStatic(Example.class, "exceptionHandler",
+MethodType.methodType(int.class, IllegalArgumentException.class, String.class));
+MethodHandle wrapped = MethodHandles.catchException(methodHandle, IllegalArgumentException.class, handler);
+
+System.out.println(wrapped.invoke("valid")); // outputs "1"
+System.out.println(wrapped.invoke("invalid")); // outputs "0"
+```
+
+### `finally` wrapper
+
+The `MethodHandles.tryFinally` method works similarly, but instead of an exception handler, it wraps an original method adding a try-finally block.
+
+Let's say we have a separate method `cleanupMethod` containing cleanup logic. The return type of this method must be the same as the original method's return type. 
+It must accept a `Throwable` followed by the resulting value coming from the original method, followed by all the arguments.
+
+```java
+public static int cleanupMethod(IllegalArgumentException e, int result, String argument) {
```

Comment:
What's the point of the `IllegalArgumentException` argument here? You are later creating a `tryFinally` handle on `wrapped` where the `IllegalArgumentException` is handled already so there cannot be an `IllegalArgumentException` passed to `cleanupMethod` here.
If we print `e` in `cleanupMethod` and run the above example, it will be `null` even or the invalid argument.
Also, if any other exception was thrown from the method, we would get a `ClassCastException` from the attempt of casting the thrown exception to an `IllegalArgumentException`.